### ISSUE TITLE
unrecoverable i/o timeout (remote_syslog_0.17)

### ISSUE THREAD

**Issue by t0m** [papertrail/remote_syslog2#176]
Hi, two of our machines stopped reporting to papertrail on the 15th around noon EST. Both were sending logs to logs3.papertrailapp.com.  The strace from one of the remote_syslog2 processes showed the following error repeating every ten seconds:

```
read(4, "2016-10-17 12:47:47 ERROR  remote_syslog.go:129 Syslog error: write tcp 173.247.107.220:17159: i/o timeout\n", 32768) = 107
write(5, "2016-10-17 12:47:47 ERROR  remote_syslog.go:129 Syslog error: write tcp 173.247.107.220:17159: i/o timeout\n", 107) = 107
read(4, "2016-10-17 12:47:57 ERROR  remote_syslog.go:129 Syslog error: write tcp 173.247.107.220:17159: i/o timeout\n", 32768) = 107
write(5, "2016-10-17 12:47:57 ERROR  remote_syslog.go:129 Syslog error: write tcp 173.247.107.220:17159: i/o timeout\n", 107) = 107
```

The process on the other machine was stuck on `read(4,` and the gdb backtrace was:

```
(gdb) bt
#0  syscall.Syscall () at /usr/local/go/src/pkg/syscall/asm_linux_amd64.s:29
#1  0x000000000056f615 in syscall.read (fd=27, p=..., n=0, err=...) at /usr/local/go/src/pkg/syscall/zsyscall_linux_amd64.go:838
#2  0x0000000000000000 in ?? ()
(gdb) 
```

Logging resumed just fine after we restarted the processes. It looks like the reads and writes are getting ENOTCONN (107) errors, is this something that remote_syslog2 can detect and recover from?

**Comment by snorecone** [comment]
@t0m this is very helpful, thank you. In the `read` backtrace, I don't suppose you know what fd=27 was or whether or not it was closed or truncated? This one seems related to an issue that I'm just beginning to explore now related to log rotation.

Regarding the `write tcp 173.247.107.220:17159` error, I'm sure it's something that should be recovered from.

**Comment by t0m** [comment]
@snorecone I didn't think to check it out when I was on the machine but just in case, for the currently running remote_syslog2 process, fds 23 - 36 are all anon_inodes according to lsof

```
remote_sy 5512 root   23r  0000     0,11          0     7820 anon_inode
remote_sy 5512 root   24r  0000     0,11          0     7820 anon_inode
remote_sy 5512 root   25r  0000     0,11          0     7820 anon_inode
remote_sy 5512 root   26r  0000     0,11          0     7820 anon_inode
remote_sy 5512 root   27r  0000     0,11          0     7820 anon_inode
remote_sy 5512 root   28r  0000     0,11          0     7820 anon_inode
remote_sy 5512 root   29r  0000     0,11          0     7820 anon_inode
remote_sy 5512 root   30r  0000     0,11          0     7820 anon_inode
remote_sy 5512 root   31r  0000     0,11          0     7820 anon_inode
remote_sy 5512 root   32r  0000     0,11          0     7820 anon_inode
remote_sy 5512 root   33r  0000     0,11          0     7820 anon_inode
remote_sy 5512 root   34r  0000     0,11          0     7820 anon_inode
remote_sy 5512 root   35r  0000     0,11          0     7820 anon_inode
remote_sy 5512 root   36r  0000     0,11          0     7820 anon_inode
```

None of the log files were truncated, they all have data from before and after the issue occurred. I'm not sure whether the fd was closed though.

**Comment by azell** [comment]
I have seen similar issues in logspout (https://github.com/gliderlabs/logspout/issues/108 and https://github.com/gliderlabs/logspout/issues/215) where server rotation (e.g. the DNS entry for the syslog server is set to a new IP with the previous IP shut down) is only fixed by killing the logspout process.

**Comment by snorecone** [comment]
Hi @t0m I believe this should be fixed with the latest version: [v0.19](https://github.com/papertrail/remote_syslog2/releases/tag/v0.19). Can you try it out?

**Comment by tterrace** [comment]
Thanks! I've updated and I haven't seen any issues since. If it does happen again I'll comment but I think you can close this out.
